### PR TITLE
[Proposal] Don't allow empty, space or newline only string in tag pattern.

### PR DESCRIPTION
**Pull request by ra1028**
Hi✋ 
This is tiny proposal.
I think that tag pattern shouldn't allow empty, space or newline only string.

In validation, returns `false` asap if found string that empty, space or newline only in the tag pattern.

### CODE REVIEW

File: Sources/Puree/TagPattern.swift [hunk 1]
```diff
@@ -19,10 +19,18 @@ public struct TagPattern {
...
-        return wildcards.count <= 1
+    private static func isValidPattern(_ pattern: String) -> Bool {
+        enum ValidationError: Error {
+            case invalidElement
+        }
+        
+        let elements = pattern.split(separator: separator, omittingEmptySubsequences: false)
+        
+        let wildcards = try? elements.filter { element in
+            guard !element.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { throw ValidationError.invalidElement }
```
Comment (giginet) [inline Sources/Puree/TagPattern.swift]: I think throwing exception doesn't make sense.
You should just return `false` instead of throwing exceptions.

**Review by giginet — COMMENTED** [review]
@ra1028 Hi 🙌 

Thanks for your contribution 👍 
Your proposal seems good. I'll merge this after changing more clearly code.

I tried this implementation. How about this?

```swift
private static func isValidPattern(_ pattern: String) -> Bool {
    let patternElements = pattern.split(separator: separator, omittingEmptySubsequences: false)
    let isValidWildcardCount = patternElements.filter { $0 == allWildcard || $0 == wildcard }.count <= 1
    let containsInvalidElement = patternElements.contains { $0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
    return isValidWildcardCount && !containsInvalidElement
}
```

**Comment by ra1028** [comment]
I think that's fine too.
However, interrupt processing asap by throwing error when found an invalid element in my codes.
Run `Sequence.filter` and` Sequence.contains` separately is obviously cause the loss of performance.
I referred to `Sequence.first(where:)` of Swift stdlib.
https://github.com/apple/swift/blob/master/stdlib/public/core/Sequence.swift#L991

File: Sources/Puree/TagPattern.swift (hunk 1 above)
Comment (slightair) [reply Sources/Puree/TagPattern.swift]: +1
and use assertionFailure

Comment (ra1028) [reply Sources/Puree/TagPattern.swift]: > I think that's fine too.
> However, interrupt processing asap by throwing error when found an invalid element in my codes.
> Run Sequence.filter andSequence.contains separately is obviously cause the loss of performance.
> I referred to Sequence.first(where:) of Swift stdlib.
> https://github.com/apple/swift/blob/master/stdlib/public/core/Sequence.swift#L991

What do you think about this?

**Comment by giginet** [comment]
> Run Sequence.filter andSequence.contains separately is obviously cause the loss of performance.

You are right about this. I tried benchmark, your code is faster.
https://gist.github.com/giginet/37dbae11c9a157da776d7faf44d0f8df

However, I think it is a bit hard to understand around here.

https://github.com/ra1028/Puree-Swift/blob/6653c1e620285841a1961d41234db78cb22e43a4/Sources/Puree/TagPattern.swift#L29-L33

We have to focus keeping readbility rather than performance.
Do you think how about this?

**Comment by ra1028** [comment]
Thanks @giginet 
Off course, will up to your coding conventions 👌

**Comment by ra1028** [comment]
@giginet 
I also added a fix for CI, is it okay?
https://github.com/ra1028/Puree-Swift/blob/6b037628c1a0f59c98b39c0c2053f3316932da04/.travis.yml#L8

**Review by giginet — APPROVED** [review]
Thanks @ra1028.

I'll release newer version soon.

**Pull request merged**